src/rules/ec2/EC2IMDSv2.ts

```diff
@@ -0,0 +1,138 @@
+              instanceLaunchTemplate.launchTemplateName,
+              instanceLaunchTemplate.launchTemplateId
+            ) &&
+            hasHttpTokens(child)
+          ) {
+            return NagRuleCompliance.COMPLIANT;
+          }
+        }
+      }
+      return NagRuleCompliance.NON_COMPLIANT;
+    } else if (node instanceof CfnAutoScalingGroup) {
+      if (node.launchTemplate) {
+        let nodeLaunchTemplate = Stack.of(node).resolve(node.launchTemplate);
+
+        for (const child of Stack.of(node).node.findAll()) {
+          if (child instanceof CfnLaunchTemplate) {
+            if (
+              isMatchingLaunchTemplate(
+                child,
+                nodeLaunchTemplate.launchTemplateName,
+                nodeLaunchTemplate.launchTemplateId
+              ) &&
+              hasHttpTokens(child)
+            ) {
+              return NagRuleCompliance.COMPLIANT;
+            }
+          }
+        }
+      } // end of check on launchTemplate
+
+      // an ASG may use a a LaunchConfiguration
+      else if (node.launchConfigurationName) {
+        let nodeLaunchConfigurationName = Stack.of(node).resolve(
+          node.launchConfigurationName
+        );
+
+        for (const child of Stack.of(node).node.findAll()) {
+          if (child instanceof CfnLaunchConfiguration) {
+            if (
+              child.launchConfigurationName === nodeLaunchConfigurationName &&
+              launchConfigurationhasTokens(child)
+            ) {
+              return NagRuleCompliance.COMPLIANT;
+            }
+          }
+        }
+      } /// end of checking LaunchConfiguration
+
+      return NagRuleCompliance.NON_COMPLIANT;
+    } else {
+      return NagRuleCompliance.NOT_APPLICABLE;
+    }
+  },
+  'name',
+  { value: parse(__filename).name }
+);
+
+function isMatchingLaunchTemplate(
+  node: CfnLaunchTemplate,
+  launchTemplateName?: string | undefined,
+  launchTemplateId?: string | undefined
+): boolean {
+  return (
+    launchTemplateName === node.launchTemplateName ||
+    launchTemplateId === NagRules.resolveResourceFromInstrinsic(node, node.ref)
+  );
+}
+
+function hasHttpTokens(node: CfnLaunchTemplate): boolean {
+  const launchTemplateData: CfnLaunchTemplate.LaunchTemplateDataProperty =
+    Stack.of(node).resolve(node.launchTemplateData);
+  const meta = Stack.of(node).resolve(
+    launchTemplateData.metadataOptions
+  ) as CfnLaunchTemplate.MetadataOptionsProperty;
+
+  if (meta == undefined) {
+    return false;
+  }
+
+  if (meta.httpTokens === 'required') {
+    return true;
+  }
+  return false;
+}
+
+function launchConfigurationhasTokens(node: CfnLaunchConfiguration): boolean {
+  if (node.metadataOptions != undefined) {
+    const meta: CfnLaunchTemplate.MetadataOptionsProperty = Stack.of(
+      node
+    ).resolve(
+      node.metadataOptions
+    ) as CfnLaunchTemplate.MetadataOptionsProperty;
+
+    if (meta.httpTokens === 'required') {
+      return true;
+    }
+  }
+  return false;
+}
```

Comment:
This can be simplified

```suggestion
function hasHttpTokens(node: CfnLaunchTemplate | CfnLaunchConfiguration): boolean {
  const metaLocation = node instanceof CfnLaunchTemplate ? Stack.of(node).resolve(node.launchTemplateData) : node.metadataOptions;
  const meta = Stack.of(node).resolve(metaLocation.metadataOptions)
  return Stack.of(node).resolve(meta?.httpTokens) === 'required'
}
```